rfc/rfc-51/rfc-51.md

```diff
@@ -0,0 +1,234 @@
+
+
+## Background
+
+In some use cases where hudi tables is used as streaing source, We want to be aware every records' changing in one commit exactly. In a certain commit or snapshot, which records are inserted, which records are deleted, and which records are updated. Even for the updated records, both the old value before updated and the new value after updated are needed.
+
+To implement this capacity, we have to upgrade the write and read parts. Let hudi can figure out the changing data when read. And in some cases, writing the extra data to help querying the changing data if necessary.
+
+## Scenario Definition
+
+Here use a simple case to explain the CDC.
+
+![](scenario-definition.jpg)
+
+Here we follow the debezium output format, there are four columns shown:
+- op: the operation of this record;
+- ts_ms: the timestamp;
+- source: source information such as the name of database and table. **Maybe we don't need this column in Hudi**;
+- before: the previous image before this operation;
+- after: the current image after this operation;
+
+`op` column has three enum values:
+- i: represent `insert`; when `op` is `i`, `before` is always null;
+- u: represent `update`; when `op` is `u`, both `before` and `after` don't be null;
+- d: represent `delete`; when `op` is `d`, `after` is always null;
+
+Notice:
+Here the illustration ignores all the metadata columns in `before` and `after` columns like `_hoodie_commit_time`.
+
+## Goals
+
+1. Support row-level CDC records generation and persistence;
+2. Support both MOR and COW tables;
+3. Support all the write operations;
+4. Support Spark DataFrame/SQL/Streaming Query;
+
+## Implementation
+### CDC Architecture
+
+![](arch.jpg)
+
+Notice:
+Other operations like `Compact`, `Clean`, `Index` do not write/change any data. So we don't need to consider in CDC scenario.
+ 
+### Points to be upgraded
+
+![](points.jpg)
+
+### Config Definitions
+
+|  | default |  |
+| --- | --- | --- |
+| hoodie.table.cdf.enabled | false | `true` represents the table to be used for CDC queries and will write cdc data if needed. |
+|  |  |  |
+| hoodie.datasource.read.cdc.enabled | false | if true, return the CDC data. |
+| hoodie.datasource.read.start.timestamp | - | requried. |
+| hoodie.datasource.read.end.timestamp | - | optional. |
+
+
+### Write
+
+Hoodie writes data by `HoodieWriteHandle`.
+We notice that only `HoodieMergeHandle` and it's subclasses will receive both the old record and the new-coming record at the same time, merge and write.
+So we will add a `LogFormatWriter` in these classes. If there is CDC data need to be written out, then call this writer to write out a log file which consist of, maybe `CDCBlock`.
+The CDC log file will be placed in the same position as the base files and other log files, so that the clean service can clean up them without extra work.
+
+The directory of the CDC file is`tablePath/.cdc/`. The file structure is like:
```

Comment:
> I thought it has been solved at the last discussion


I remember that the last discussion conclusion is:
- for COW table, derives the cdc details on the fly when reading
- for MOR table, make a new configuration option for deciding whether to encode the changes logs in separate block

Correct me if i'm wrong. The separate .cdc  files and directory was dropped on the very first decision, did i remember it wrong ? And i want to say again that each design decision is IMPORTANT because it would affect the file layout, we should put enough thinkings on the trade off and whether the design is production ready and robust enough.

Reply:
>The separate .cdc files and directory was dropped on the very first decision, did i remember it wrong ?

That is my understanding as well. 

> for COW table, derives the cdc details on the fly when reading
this part I think we still were debating whether to materialize in a log file or compute on the fly. I ll go re-read this again.

Reply:
>each design decision is IMPORTANT because it would affect the file layout, 

+100 and they are all interdependent. I think we should explicitly track a comment with list of decisions and outcomes.